vllm/v1/attention/backends/flashinfer.py

```diff
@@ -0,0 +1,497 @@
+# SPDX-License-Identifier: Apache-2.0
+"""Attention layer with FlashAttention."""
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Tuple, Type
+
+try:
+    from flashinfer import (BatchPrefillWithPagedKVCacheWrapper,
+                            MultiLevelCascadeAttentionWrapper)
+    FLASHINFER_WORKSPACE_BUFFER_SIZE = 256 * 1024 * 1024
+except ImportError:
+    # Avoid turning these types into variables during type checking
+    if not TYPE_CHECKING:
+        BatchPrefillWithPagedKVCacheWrapper = None
+        MultiLevelCascadeAttentionWrapper = None
+    FLASHINFER_WORKSPACE_BUFFER_SIZE = 0
+
+import numpy as np
+import torch
+
+from vllm.attention.backends.abstract import (AttentionImpl, AttentionLayer,
+                                              AttentionMetadata, AttentionType)
+from vllm.attention.layer import Attention
+from vllm.config import VllmConfig, get_current_vllm_config
+from vllm.logger import init_logger
+from vllm.utils import cdiv
+
+
+logger = init_logger(__name__)
+
+
+class FlashInferBackend:
```

Comment:
afaict all attention backend implementation should be a subclass of `AttentionBackend`, as there are function like `get_name` for metadata